controllers/imapmon.go

```diff
@@ -0,0 +1,184 @@
+package controllers
+
+/* TODO:
+*		 - Have a counter per config for number of consecutive login errors and backoff (e.g if supplied creds are incorrect)
+*		 - Have a DB field "last_login_error" if last login failed
+*		 - DB counter for non-campaign emails that the admin should investigate
+*		 - Add field to User for numner of non-campaign emails reported
+ */
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/gophish/gophish/logger"
+
+	"github.com/glennzw/eazye"
+	"github.com/gophish/gophish/config"
+	"github.com/gophish/gophish/models"
+)
+
+// Pattern for GoPhish emails e.g ?rid=AbC123
+var goPhishRegex = regexp.MustCompile("(\\?rid=[A-Za-z0-9]{7})")
+
+// ImapMonitor is a worker that monitors IMAP serverd for reported campaign emails
+type ImapMonitor struct {
+	cancel    func()
+	reportURL string
+	checkFreq int
+}
+
+// PeriodicIMAPMonitor will periodically check the database for
+func PeriodicIMAPMonitor(ctx context.Context, im *ImapMonitor) {
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			imaps, err := models.GetEnabledIMAPs()
+			if err != nil {
+				log.Error(err)
+				break
+			}
+			var wg sync.WaitGroup
+			wg.Add(len(imaps))
+			for _, i := range imaps {
+				//We launch a goroutine for each IMAP account. In the future should use some more sophisticated worker/queues
+				go checkForNewEmails(i, &wg, im.reportURL)
+			}
+			wg.Wait()
+			time.Sleep(time.Duration(im.checkFreq) * time.Second)
+		}
+	}
+
+}
+
+// NewImapMonitor returns a new instanfe of the ImapMonitor
+//func NewImapMonitor(configP config.PhishServer, configA *config.Config) *ImapMonitor {
+func NewImapMonitor(config *config.Config) *ImapMonitor {
+	// Make sure database connection exists. Not sure why I have to do this here, but
+	//  otherwise db is <nil> when calling models.GetEnabledIMAPs()
+	err := models.Setup(config)
+	if err != nil {
+		log.Fatal(err)
+	}
+	reportURL := "http://" + config.PhishConf.ListenURL
+	if config.PhishConf.UseTLS {
+		reportURL = "https://" + config.PhishConf.ListenURL
+	}
+
+	im := &ImapMonitor{
+		reportURL: reportURL,
+		checkFreq: config.IMAPFreq,
+	}
+	return im
+}
+
+// Start launches the IMAP campaign monitor
+func (im *ImapMonitor) Start() error {
+	log.Info("Starting IMAP monitor")
+	ctx, cancel := context.WithCancel(context.Background()) //ctx is the derivedContext
+	im.cancel = cancel
+	go PeriodicIMAPMonitor(ctx, im)
```

Comment:
What do you think about moving `PeriodicIMAPMonitor` to something like `im.start(ctx)` or some other name? Right now, it's not clear what use case we'd have that would require calling the exported `PeriodicIMAPMonitor` directly from outside the package.

Reply:
Done. Removed `PeriodicIMAPMonitor` and added `im.start(ctx)`